### PR TITLE
PF use raw ecal for PF hadron calibration

### PR DESCRIPTION
**Pull request by hatakeyamak**
With this change, when the PF hadron calibration is called in PFAlgo, the raw ECAL energy will be used as inputs to pull the PF hadron calibration factor. Note that, when this calibration is derived, raw ECAL energy is used as inputs, and it allows the PF ECAL cluster calibration and PF hadron cluster calibration to be orthogonal.

The expected change (evaluated in a few MC samples):
+ effect in jet energy scale <1%, negligible in resolution (evaluated in QCD flat without PU) [1, page-4] [2] [3]
+ up to 10% reduction in PF photon energy and up to 1% reduction in PF neutral hadron energy (evaluated in neutrino gun MC with PU, L1 offset like observables) [1, page-6]

[1] https://indico.cern.ch/event/786078/contributions/3274894/attachments/1778982/2893356/PF_hadron_calib_20190115.pdf
[2] https://indico.cern.ch/event/777968/contributions/3255724/attachments/1773152/2882370/juska_pf_20181812.pdf
[3] https://indico.cern.ch/event/786078/contributions/3274894/attachments/1778982/2893550/juska_pf_20190115.pdf

@zdemirag @VinInn @cbernet @jpata @seemasharmafnal

### CODE REVIEW

**Comment by hatakeyamak** [comment]
I am not too familiar with checking outcomes of Jenkins tests... 

Can I assume that this will be reviewed when people become available (in other words, there is no pending requests to me.) ?

Thank you.

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 1]
```diff
@@ -2895,22 +2903,27 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
...
       cand.addElementInBlock( blockref, sortedTracks.begin()->second) ;
-    }
-
-
-  }
 
-   // end loop on hcal element iHcal= hcalIs[i]
+      if ( fabs(eclusterref->energy() - sqrt(std::get<1>(ecalSatellite.second).Mag2()))>1e-3 || fabs(eclusterref->correctedEnergy() - ecalClusterEnergyCalibrated)>1e-3 )
+	edm::LogWarning("PFAlgo:processBlock") << "ecalCluster vs ecalSatellites look inconsistent (eCluster E, calibE, ecalSatellite E, calib E): "
+					       << eclusterref->energy() << " " << eclusterref->correctedEnergy() << " "
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: What are the implications on the ecalSatellites data (or what follows from them)?
If the data is usable, this warning better be replaced by a LogInfo or put behind ```if (debug_```.
If the data is completely unusable and/or if this simply is not supposed to happen (e.g. indicates a bug in the code), perhaps even an exception is a better solution.
@hatakeyamak  please clarify

In my tests I did not see any instance of this happening out of about 7K events with 2018-like pileup.
So, at least from the point of view of frequency, this is OK to stay as a warning (as it does no harm so far).

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc [hunk 2]
```diff
@@ -2999,7 +3012,8 @@ void PFAlgo::processBlock( const reco::PFBlockRef& blockref,
       reco::PFClusterRef eclusterRef = elements[iEcal].clusterRef();
       assert( !eclusterRef.isNull() );
 
-      double ecalEnergy = eclusterRef->correctedEnergy();
+      // KH: use raw ECAL energy for PF hadron calibration. 
+      double ecalEnergy = eclusterRef->energy(); // ecalEnergy = eclusterRef->correctedEnergy();
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFAlgo.cc]: is the commented part of the code needed?
IIUC, there is no single place to make a switch from raw to calibrated.
So, the commented out part here can't be useful as a quick check to switch to the calibrated value.
Please drop the commented out ```// ecalEnergy = eclusterRef->correctedEnergy();``` or add some note why this is still useful here.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25703	9d6dd5f
- code changes are in line with the PR description: the pfHadron energy calibration  is now consistently using ECAL raw inputs
    - given that this is more correct from first principles, the following observations from the change are likely to be all considered as in the right direction  
- jenkins tests pass and comparisons with the baseline show mostly minor differences starting from the PF candidates and propagating downstream
- local tests with more events confirm this behavior and also appear in line with what was shown in the PF group meeting (all changes are rather modest or small)
    - there is a smaller number of pf photons and neutral hadrons
    - additionally, the number of pf charged hadrons also goes down and the number of pf muons goes up
    - it looks like the overall sumEt is reducing by a fraction of one %.
    - other objects with significant changes in counts/distributions:
        - hps pi0 yields are down by a few %
        - lostTracks in miniAOD are up by a few % (mostly at low pt, below 1 GeV). My pre-expectations were that the tracks remain used; however, given the reduction even in the number of charged hadrons, quite a few really get rejected.

In event-by event comparison in workflow 136.862 ( EGamma2018B), in 200 events:
- pf sumEt is mostly reduced: up in 2 events and down in 198 events
    - all of the 198 have the reduction in sumEt below 1%; 169 is above 0.1%; and 7 event above 0.5%

Some "representative" observations, here in ZEE events with PU (red is with this PR):
pf muons eta
all_sign1064vsorig_zee13tevpu2018wf11025p0c_recopfcandidates_particleflow__reco_obj_eta49
pf photon eta (counts drop, as expected)
all_sign1064vsorig_zee13tevpu2018wf11025p0c_recopfcandidates_particleflow__reco_obj_eta56
hps pi0 follow
all_sign1064vsorig_zee13tevpu2018wf11025p0c_recorecotaupizeros_hpspftauproducer_pizeros_reco_obj_eta
lost tracks 
all_mini_sign1064vsorig_zee13tevpu2018wf11025p0c_log10patpackedcandidates_losttracks__reco_obj_pt
all_mini_sign1064vsorig_zee13tevpu2018wf11025p0c_patpackedcandidates_losttracks__reco_obj_eta

there is ever-so-small broadening (perhaps not stat significant; however this picture repeats in dijet and ZMM tests) in raw pfMet wrt GEN MET. This is probably fixable with type1 corrections, but that would need a new JEC.
wf11025_genmet

jet response in the dijet sample (flat with PU) is down a bit (as expected)
wf11071_ak4chs_resp_e

**Comment by fabiocos** [comment]
+1

File: RecoParticleFlow/PFProducer/src/PFAlgo.cc (hunk 2 above)
Comment (hatakeyamak) [reply RecoParticleFlow/PFProducer/src/PFAlgo.cc]: OK. I will clean up when I touch PFAlgo next time.

**Pull request merged**